README.md
# Image resizer

**Created as part of [inspishop][link-inspishop] e-commerce platform by [inspirum][link-inspirum] team.**

[![Build Status][ico-travis]][link-travis]
[![Software License][ico-license]][link-licence]
[![Docker Version][ico-docker-hub-version]][link-docker-hub]
[![Docker Pulls][ico-docker-hub-downloads]][link-docker-hub]
[![Docker Image Size][ico-docker-hub-size]][link-docker-hub]

On-demand image resizing, format converting and size optimization with the best CLI tools. 
Prepared to be used as Docker image with cloud storage as Amazon S3.

- image [resizing](#resizing), [format converting](#format-encoding) and [optimization](#optimization)
- support for **jpeg**, **png**, **webp**, **bmp**, **gif**, **tiff**, **svg** image formats
- file-cache for resized images on local disk *(temporary cache)* and cloud storage *(persistent cache)*


## Usage example


### Resizing

Application handle route with dynamic [**template**](#template) (parameters divided by dash) and **filepath** parameters.
```
GET /image/{template}/{filepath}
```


#### Template

Options for resizing are specified as a dash delimited list of parameters, which can be supplied in any order. 
Duplicate parameters overwrite last values.

| Name | Template | Example | Description  |
| :--- | :--- | :--- | :--- |
| width | `w[X]` | `w200` | Resize to 200px width with proportional height |
| height| `h[X]` | `h100` | Resize to 100px height with proportional width |
| ratio | `[X]x[X]` | `1x2` |  Resize to ratio to 1:2 *(width : height)* |
| crop | `crop` | `crop` | Image fill whole canvas (no addition background added) |
| upscale | `upscale` | `upscale` | Image can be larger than original |

There are special template string `original`. This will return optimized image with original dimensions.
It is helpful to show image with maximal resolution without knowing its dimension.

When there are set both `height` and `width` then `ratio` parameter will be ignored.

If output image has a different ratio than an original, image will be centered on transparent canvas (**jpg** has white background because it doesn't support transparent color).


#### Examples

> Values are for an image with dimension 200x400px (width : height).

| Template string | Output dimension |
| :--- | :--- | 
| `original` | 200x400 |
| `h100` | 50x100 |
| `w100` |  100x200 |
| `w50-w200` |  100x200 |
| `w100-h300` | 100x300 |
| `w200-1x1` | 200x200 |
| `w200-1x3` | 200x600 |
| `w100-h300-2x1` | 100x300 |
| `w1000-1x1` | 1000x1000 |
| `w1000-1x1-crop` | 200x200 |
| `w1000-1x1-crop-upscale` | 1000x1000 |


#### Width & height

> Examples use **jpg** format because canvas background is white (for better visibility)

| `h200` | `w200` | `w188-h233` |
| :---: | :---: | :---: |
| *(300x200)* | *(200x134)*  | *(188x233)* |
| ![h200](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/sizes/h200.jpg) | ![w200](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/sizes/w200.jpg) | ![w188-h233](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/sizes/w188-h233.jpg) |


#### Ratio


| `h200-1x1` | `h200-1x2` | `h200-2x1` |
| :---: | :---: | :---: |
| *(200x200)* | *(100x200)*  | *(400x200)* |
| ![h200-1x1](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/ratio/h200-1x1.jpg) | ![h200-1x2](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/ratio/h200-1x2.jpg) | ![h200-121](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/ratio/h200-2x1.jpg) |


#### Crop

> `crop` parameter will crop image to fill whole canvas 

| `h200-1x1-crop` | `h200-1x2-crop` | `h200-2x1-crop` |
| :---: | :---: | :---: |
| *(200x200)* | *(100x200)*  | *(400x200)* |
| ![h200-1x1-crop](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/ratio/h200-1x1-crop.jpg) | ![h200-1x2-crop](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/ratio/h200-1x2-crop.jpg) | ![h200-121-crop](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/ratio/h200-2x1-crop.jpg) |


#### Upscale

> By default, image can't up-size (be larger than original) 

| `original` | `w250-h300` | `w250-h300-upscale` | `w250-h300-crop` | `w250-h300-crop-upscale` |
| :---: | :---: | :---: | :---: | :---: |
| *(200x200)* | *(250x300)*  | *(250x300)* | *(167x200)* | *(250x300)*  |
| ![original](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/upsize/original.jpg) | ![w250-h300](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/upsize/w250-h300.jpg) | ![w250-h300-upscale](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/upsize/w250-h300-upscale.jpg) | ![w250-h300-crop](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/upsize/w250-h300-crop.jpg) | ![w250-h300-crop-upscale](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/upsize/w250-h300-crop-upscale.jpg) |

| `original` | `w500-h300-crop` |`w500-h300-crop` | `w500-h300-crop-upscale` | 
| :---: | :---: | :---: | :---: |
| *(200x200)* |  *(500x300)* | *(200x120)* | *(500x300)* |
| ![original](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/upsize/original.jpg) | ![w500-h300](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/upsize/w500-h300.jpg) | ![w500-h300-crop](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/upsize/w500-h300-crop.jpg) | ![w5w500-h300-crop-upscale](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/upsize/w500-h300-crop-upscale.jpg) |


### Format encoding

Image can be encoded to any supported format simple by changing file extension in URL.

Original image extension is specified in `original` query parameter.

> This is typically used to encode image as web-friendly **webp** format
```
http://localhost:3000/image/custom-w400-1x1-upscale/data/products/0/1/background.webp?original=png
```


### Optimization

The package will use these optimizers if they are present on your system:

- [pngquant 2](https://pngquant.org)
- [jpegoptim](http://freecode.com/projects/jpegoptim)
- [gifsicle](http://www.lcdf.org/gifsicle)
- [svgcleaner](https://github.com/RazrFalcon/svgcleaner)
- [cwebp](https://developers.google.com/speed/webp/docs/cwebp)

By setting environment variable you can change default path to binary file or default arguments. 

| Environment variable | Default value |
| :--- | :--- | 
| `CMD_OPTIMIZER_PNG` | `pngquant --force --ext .png --skip-if-larger --quality 0-75 --speed 4 --strip --` |
| `CMD_OPTIMIZER_JPG` | `jpegoptim --force --strip-all --max 75 --quiet --all-progressive` |
| `CMD_OPTIMIZER_GIF` | `gifsicle --batch --optimize=3` |
| `CMD_OPTIMIZER_SVG` | `svgcleaner` |
| `CMD_OPTIMIZER_WEBP` | `cwebp -m 6 -pass 10 -mt -q 75 -quiet` |


#### Examples

| 64 kB | 28 kB *(43 %)* | 22 kB *(34 %)* |
| :---: | :---: | :---: |
| ![original](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/optimize/jpg_original.jpg) | ![optimized](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/optimize/jpg_optimized.jpg) | ![optimized](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/optimize/jpg_optimized.webp) |

| 421 kB | 91 kB *(21 %)* | 23 kB *(5 %)* |
| :---: | :---: | :---: |
| ![original](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/optimize/png_original.png) | ![optimized](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/optimize/png_optimized.png) | ![optimized](https://github.com/inspirum/assets/raw/master/image-resizer-go/img/optimize/png_optimized.webp) |


## System requirements


## Installation


## Testing


## Contributing

Please see [CONTRIBUTING][link-contributing] and [CODE_OF_CONDUCT][link-code-of-conduct] for details.


## Security

If you discover any security related issues, please email tomas.novotny@inspirum.cz instead of using the issue tracker.


## Credits

- [Tomáš Novotný](https://github.com/tomas-novotny)
- [All Contributors][link-contributors]


## License

The MIT License (MIT). Please see [License File][link-licence] for more information.


[ico-license]:              https://img.shields.io/github/license/inspirum/image-resizer-go.svg?style=flat-square&colorB=blue
[ico-travis]:               https://img.shields.io/travis/inspirum/image-resizer-go/master.svg?branch=master&style=flat-square
[ico-docker-hub-version]:   https://img.shields.io/docker/v/inspirum/image-resizer?sort=semver&style=flat-square&label=docker%20tag
[ico-docker-hub-size]:      https://img.shields.io/docker/image-size/inspirum/image-resizer?style=flat-square&label=docker%20image%20size
[ico-docker-hub-downloads]: https://img.shields.io/docker/pulls/inspirum/image-resizer?style=flat-square


[link-author]:              https://github.com/inspirum
[link-contributors]:        https://github.com/inspirum/image-resizer-go/contributors
[link-licence]:             ./LICENSE.md
[link-changelog]:           ./CHANGELOG.md
[link-contributing]:        ./docs/CONTRIBUTING.md
[link-code-of-conduct]:     /docs/CODE_OF_CONDUCT.md
[link-travis]:              https://travis-ci.org/inspirum/image-resizer-go
[link-docker-hub]:          https://hub.docker.com/repository/docker/inspirum/image-resizer/tags?page=1
[link-inspishop]:           https://www.inspishop.cz/
[link-inspirum]:            https://www.inspirum.cz/


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased](https://github.com/inspirum/image-resizer-go/compare/v0.1.0...master)


## v0.1.2 (2020-07-13)
### Fixes
- Fixes cloud file-cache


## v0.1.1 (2020-07-13)
### Fixes
- Prints error for resized image file-cache  


## v0.1.0 (2020-07-13)
### Added
- Added basic functionality

resizer_test.go
package imageresizer

import (
	"testing"
)

func TestTemplate(t *testing.T) {

	template := NewTemplate("custom-w200")

	if template.width != 200 {
		t.Error("With failed")
	}

	if template.height != 0 {
		t.Error("Height failed")
	}
}


s3.go
package main

import (
	"fmt"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"io"
	"os"
	"path/filepath"
	"time"
)

type s3client struct {
	s3     *s3.S3
	bucket string
}

func NewS3Client(endpoint string, bucket string, key string, secret string, region string) *s3client {
	s, err := session.NewSession(&aws.Config{
		Endpoint:    aws.String(endpoint),
		Region:      aws.String(region),
		Credentials: credentials.NewStaticCredentials(key, secret, ""),
	})

	if err != nil {
		fmt.Printf("Unable create New S3 s %v\n", err)
		os.Exit(1)
	}

	return &s3client{s3.New(s), bucket}
}

func (c *s3client) DownloadFile(path string, outputPath string) (f *os.File, err error) {
	f, _, err = c.DownloadFileWithModTime(path, outputPath)

	return
}

func (c *s3client) DownloadFileWithModTime(path string, outputPath string) (f *os.File, t *time.Time, err error) {
	resp, err := c.getContent(path)

	if err != nil {
		return nil, nil, err
	}

	t = resp.LastModified
	defer resp.Body.Close()

	if outputPath == "" {
		f, err = CreateTempFileFromReader(resp.Body, filepath.Ext(path))
	} else {
		_, err = WriteFileFromReader(outputPath, resp.Body)
		if err != nil {
			return
		}
		f, err = os.Open(outputPath)
	}

	return
}

func (c *s3client) UploadContentReaderIfNewer(path string, time time.Time, content io.ReadSeeker) (bool, error) {
	resp, err := c.headContent(path)

	if err == nil && resp.LastModified.After(time) {
		return false, nil
	}

	err = c.UploadContentReader(path, content)

	if err != nil {
		return false, err
	}

	return true, nil
}

func (c *s3client) UploadContentReader(path string, content io.ReadSeeker) error {
	mime, err := GetReaderContentType(content)
	if err != nil {
		return err
	}

	return c.putContent(path, content, mime)
}

func (c *s3client) putContent(path string, content io.ReadSeeker, mime string) (err error) {
	_, err = c.s3.PutObject(&s3.PutObjectInput{
		Key:                aws.String(path),
		Body:               content,
		Bucket:             aws.String(c.bucket),
		ContentType:        aws.String(mime),
		ContentDisposition: aws.String("attachment"),
	})

	return
}

func (c *s3client) getContent(path string) (resp *s3.GetObjectOutput, err error) {
	resp, err = c.s3.GetObject(&s3.GetObjectInput{
		Bucket: aws.String(c.bucket),
		Key:    aws.String(path),
	})

	return
}

func (c *s3client) headContent(path string) (resp *s3.HeadObjectOutput, err error) {
	resp, err = c.s3.HeadObject(&s3.HeadObjectInput{
		Bucket: aws.String(c.bucket),
		Key:    aws.String(path),
	})

	return
}


app/main.go
package main

import (
	"fmt"
	imageresizer "github.com/inspirum/image-resizer-go"
	"github.com/julienschmidt/httprouter"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"sync"
	"time"
)

type server struct {
	router           *httprouter.Router
	storage          imageresizer.Storage
	localCache       bool
	localCacheRoot   string
	cloudCache       bool
	cloudCacheRoot   string
	cacheMaxAge      int
	notFoundFilename string
}

func main() {
	routerMux := httprouter.New()
	s3Service := imageresizer.NewS3Client(
		imageresizer.GetEnv("S3_ENDPOINT", ""),
		imageresizer.GetEnv("S3_BUCKET", ""),
		imageresizer.GetEnv("S3_KEY", ""),
		imageresizer.GetEnv("S3_SECRET", ""),
		imageresizer.GetEnv("S3_REGION", ""),
	)

	s := &server{
		routerMux,
		s3Service,
		true,
		imageresizer.GetEnv("STORAGE_LOCAL_PREFIX", "./cache/"),
		imageresizer.GetEnvAsBool("STORAGE_CLOUD_ENABLED", true),
		imageresizer.GetEnv("STORAGE_CLOUD_PREFIX", ""),
		imageresizer.GetEnvAsInt("CACHE_MAX_AGE", 7200),
		imageresizer.GetEnv("NOTFOUND_FILENAME", "./static/no_image.png"),
	}

	s.router.GET("/image/:template/*filepath", s.ServeFile)

	port := imageresizer.GetEnvAsInt("PORT", 3000)
	fmt.Printf("Listening on http://localhost:%d/\n", port)
	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", port), routerMux))
}

func (s *server) ServeFile(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
	template := p.ByName("template")
	filename := p.ByName("filepath")

	logger("\n== REQ %s == %s%s?%s\n", time.Now().Format("15:04:05.000"), template, filename, r.URL.RawQuery)

	if err := imageresizer.ValidateTemplate(template); err != nil {
		s.buildErrorResponse(w, err)
		return
	}

	originalFilename := imageresizer.ReplacePathExt(filename, r)
	if err := imageresizer.ValidateFilename(originalFilename); err != nil {
		s.buildErrorResponse(w, err)
		return
	}

	resizedFilename := template + filename
	resizedFile, modTime, err := s.getResizedImageFile(resizedFilename)
	if err == nil {
		s.buildResponse(w, r, resizedFile, *modTime, resizedFilename)
		return
	}

	originalFile, err := s.getOriginalImageFile(originalFilename)
	if err != nil {
		s.buildNotFoundResponse(w, r, template)
		return
	}
	_ = originalFile.Close()
	defer os.Remove(originalFile.Name())

	resizedFile, err = s.resizeImageFile(originalFile.Name(), s.getLocalResizedPath(resizedFilename), template)
	if err != nil {
		s.buildErrorResponse(w, err)
		return
	}

	s.buildResponse(w, r, resizedFile, time.Now(), resizedFilename)
}

func (s *server) getLocalResizedPath(path string) string {
	return s.localCacheRoot + path
}

func (s *server) getCloudResizedPath(path string) string {
	return s.cloudCacheRoot + path
}

func (s *server) getResizedImageFile(path string) (f *os.File, t *time.Time, err error) {
	vLogger("Get local resized image %s\n", s.getLocalResizedPath(path))
	f, t, err = imageresizer.OpenFileWithModTime(s.getLocalResizedPath(path))

	if err != nil && s.cloudCache {
		vLogger("Get cloud resized image cloud://%s\n", s.getCloudResizedPath(path))
		f, t, err = s.storage.DownloadFileWithModTime(s.getCloudResizedPath(path), s.getLocalResizedPath(path))
	}

	return
}

func (s *server) getOriginalImageFile(path string) (*os.File, error) {
	vLogger("Get original image cloud:/%s\n", path)
	return s.storage.DownloadFile(path, "")
}

func (s *server) resizeImageFile(inputFilePath string, outputFilePath string, template string) (*os.File, error) {
	if imageresizer.ShouldNotResize(template, inputFilePath) {
		vLogger("Return original\n")
		err := imageresizer.CopyFile(inputFilePath, outputFilePath)
		if err != nil {
			return nil, err
		}
	} else {
		vLogger("Return resized\n")
		w, h, err := imageresizer.GetImageDimensions(inputFilePath)
		if err != nil {
			return nil, err
		}

		err = imageresizer.ConvertFile(inputFilePath, w, h, outputFilePath, imageresizer.NewTemplate(template))
		if err != nil {
			return nil, err
		}
	}

	err := imageresizer.OptimizeFile(outputFilePath)
	if err != nil {
		return nil, err
	}

	return os.Open(outputFilePath)
}

func (s *server) writeResizedImageFile(filePath string, outputFilePath string, wg *sync.WaitGroup) {
	if s.localCache {
		localResizedPath := s.getLocalResizedPath(outputFilePath)
		vLogger("- [async] Write local resized image %s\n", localResizedPath)
		wg.Add(1)
		go func() {
			time.Sleep(1)
			defer wg.Done()
			content, err := os.Open(filePath)
			if err != nil {
				vLogger("- [await] Error local resized image %s: %v\n", localResizedPath, err)
			}
			changed, err := imageresizer.WriteFileFromReader(localResizedPath, content)
			_ = content.Close()
			if err != nil {
				vLogger("- [await] Error local resized image %s: %v\n", localResizedPath, err)
			} else if changed {
				vLogger("- [await] Written local resized image %s\n", localResizedPath)
			} else {
				vLogger("- [await] Unchanged local resized image %s\n", localResizedPath)
			}
		}()
	}

	if s.cloudCache {
		cloudResizedPath := s.getCloudResizedPath(outputFilePath)
		vLogger("- [async] Write cloud resized image cloud://%s\n", cloudResizedPath)
		wg.Add(1)
		go func() {
			defer wg.Done()
			content, err := os.Open(filePath)
			if err != nil {
				vLogger("- [await] Error cloud resized image %s: %v\n", cloudResizedPath, err)
			}
			changed, err := s.storage.UploadContentReaderIfNewer(cloudResizedPath, time.Now().Add(-1*time.Second*time.Duration(s.cacheMaxAge)), content)
			_ = content.Close()
			if err != nil {
				vLogger("- [await] Error cloud resized image %s: %v\n", cloudResizedPath, err)
			} else if changed {
				vLogger("- [await] Written cloud resized image cloud://%s\n", cloudResizedPath)
			} else {
				vLogger("- [await] Unchanged cloud resized image cloud://%s\n", cloudResizedPath)
			}
		}()
	}
}

func (s *server) buildResponse(w http.ResponseWriter, r *http.Request, content *os.File, modTime time.Time, resizedFilePath string) {
	var wg sync.WaitGroup

	w.Header().Set("Cache-Control", fmt.Sprintf("max-age=%d, public", s.cacheMaxAge))
	http.ServeContent(w, r, resizedFilePath, modTime, content)
	_ = content.Close()

	s.writeResizedImageFile(content.Name(), resizedFilePath, &wg)

	logger("== RESP %s == %d %s\n", time.Now().Format("15:04:05.000"), http.StatusOK, resizedFilePath)

	go func() {
		wg.Wait()
		vLogger("- [await] Finish response\n")
	}()
}

func (s *server) buildNotFoundResponse(w http.ResponseWriter, r *http.Request, template string) {
	vLogger("Return 404 response\n")
	resizedFilePath := template + "/" + filepath.Base(s.notFoundFilename)
	resizedFile, _, err := s.getResizedImageFile(resizedFilePath)

	if err != nil {
		resizedFile, err = s.resizeImageFile(s.notFoundFilename, s.getLocalResizedPath(resizedFilePath), template)

		if err != nil {
			s.buildErrorResponse(w, err)
			return
		}
	}

	// TODO: add cache precondition
	w.Header().Set("Cache-Control", "max-age=60, public")
	statusCode, _ := strconv.Atoi(r.URL.Query().Get("status"))
	if statusCode < 200 || statusCode >= 500 {
		statusCode = http.StatusNotFound
	}

	w.WriteHeader(statusCode)
	_, _ = io.Copy(w, resizedFile)

	_ = resizedFile.Close()

	logger("== RESP %s == %d %s\n", time.Now().Format("15:04:05.000"), statusCode, resizedFile.Name())
}

func (s *server) buildErrorResponse(w http.ResponseWriter, err error) {
	code := http.StatusInternalServerError
	if err, ok := err.(imageresizer.HttpError); ok {
		code = err.StatusCode
	}

	w.Header().Set("Cache-Control", "max-age=60, public")
	http.Error(w, err.Error(), code)

	logger("== ERR %s == %d %s\n", time.Now().Format("15:04:05.000"), code, err.Error())
}

var verbose = imageresizer.GetEnvAsBool("VERBOSE", true)

func vLogger(format string, a ...interface{}) {
	if verbose {
		logger(format, a...)
	}
}

func logger(format string, a ...interface{}) {
	fmt.Printf(format, a...)
}


functions.go
package imageresizer

import (
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"path"
	"path/filepath"
	"strconv"
	"time"
)

type HttpError struct {
	error
	StatusCode int
}

func GetEnv(key string, fallback string) string {
	if value, ok := os.LookupEnv(key); ok {
		return value
	}

	return fallback
}

func GetEnvAsInt(key string, fallback int) int {
	if value, ok := os.LookupEnv(key); ok {
		value, _ := strconv.Atoi(value)
		return value
	}

	return fallback
}

func GetEnvAsBool(key string, fallback bool) bool {
	if value, ok := os.LookupEnv(key); ok {
		value, _ := strconv.ParseBool(value)
		return value
	}

	return fallback
}

func GetReaderContentType(c io.ReadSeeker) (mime string, err error) {
	b := make([]byte, 512)
	_, err = c.Read(b)

	if err != nil {
		return
	}

	mime = http.DetectContentType(b)

	_, err = c.Seek(0, io.SeekStart)

	return
}

func OpenFileWithModTime(p string) (*os.File, *time.Time, error) {
	f, err := os.Open(p)

	if err != nil {
		return nil, nil, err
	}

	d, err := f.Stat()
	if err != nil {
		return nil, nil, err
	}

	t := d.ModTime()

	return f, &t, nil
}

func WriteFileFromReader(p string, c io.Reader) (bool, error) {
	if _, err := os.Stat(p); !os.IsNotExist(err) {
		return false, nil
	}

	_ = os.MkdirAll(filepath.Dir(p), 0777)

	f, err := os.OpenFile(p, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0777)
	if err != nil {
		return false, err
	}

	_, err = io.Copy(f, c)
	_ = f.Close()
	if err != nil {
		return false, err
	}

	return true, nil
}

func CopyFile(src, dst string) (err error) {
	srcFile, err := os.Open(src)
	if err != nil {
		return
	}
	defer srcFile.Close()

	_ = os.MkdirAll(filepath.Dir(dst), 0777)
	dstFile, err := os.OpenFile(dst, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0777)
	if err != nil {
		return
	}
	defer dstFile.Close()

	_, err = io.Copy(dstFile, srcFile)

	return
}

func CreateTempFileFromReader(c io.Reader, ext string) (f *os.File, err error) {
	f, err = ioutil.TempFile("", "_resize_*"+ext)

	if err != nil {
		return
	}

	_, err = io.Copy(f, c)

	if err != nil {
		return
	}

	_, err = f.Seek(0, io.SeekStart)

	return
}

func ReplacePathExt(p string, r *http.Request) string {
	ext := r.URL.Query().Get("original")

	if ext == "" {
		return p
	}

	return p[0:len(p)-len(path.Ext(p))+1] + ext
}

var verbose = GetEnvAsBool("VERBOSE", true)

func logger(format string, a ...interface{}) {
	if verbose {
		fmt.Printf(format, a...)
	}
}


storage_s3.go
package imageresizer

import (
	"fmt"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"io"
	"os"
	"path/filepath"
	"time"
)

type s3client struct {
	s3     *s3.S3
	bucket string
}

func NewS3Client(endpoint string, bucket string, key string, secret string, region string) *s3client {
	s, err := session.NewSession(&aws.Config{
		Endpoint:    aws.String(endpoint),
		Region:      aws.String(region),
		Credentials: credentials.NewStaticCredentials(key, secret, ""),
	})

	if err != nil {
		fmt.Printf("Unable create New S3 s %v\n", err)
		os.Exit(1)
	}

	return &s3client{s3.New(s), bucket}
}

func (c *s3client) DownloadFile(path string, outputPath string) (f *os.File, err error) {
	f, _, err = c.DownloadFileWithModTime(path, outputPath)

	return
}

func (c *s3client) DownloadFileWithModTime(path string, outputPath string) (f *os.File, t *time.Time, err error) {
	resp, err := c.getContent(path)

	if err != nil {
		return nil, nil, err
	}

	t = resp.LastModified
	defer resp.Body.Close()

	if outputPath == "" {
		f, err = CreateTempFileFromReader(resp.Body, filepath.Ext(path))
	} else {
		_, err = WriteFileFromReader(outputPath, resp.Body)
		if err != nil {
			return
		}
		f, err = os.Open(outputPath)
	}

	return
}

func (c *s3client) UploadContentReaderIfNewer(path string, time time.Time, content io.ReadSeeker) (bool, error) {
	resp, err := c.headContent(path)

	if err == nil && resp.LastModified.After(time) {
		return false, nil
	}

	err = c.UploadContentReader(path, content)

	if err != nil {
		return false, err
	}

	return true, nil
}

func (c *s3client) UploadContentReader(path string, content io.ReadSeeker) error {
	mime, err := GetReaderContentType(content)
	if err != nil {
		return err
	}

	return c.putContent(path, content, mime)
}

func (c *s3client) putContent(path string, content io.ReadSeeker, mime string) (err error) {
	_, err = c.s3.PutObject(&s3.PutObjectInput{
		Key:                aws.String(path),
		Body:               content,
		Bucket:             aws.String(c.bucket),
		ContentType:        aws.String(mime),
		ContentDisposition: aws.String("attachment"),
	})

	return
}

func (c *s3client) getContent(path string) (resp *s3.GetObjectOutput, err error) {
	resp, err = c.s3.GetObject(&s3.GetObjectInput{
		Bucket: aws.String(c.bucket),
		Key:    aws.String(path),
	})

	return
}

func (c *s3client) headContent(path string) (resp *s3.HeadObjectOutput, err error) {
	resp, err = c.s3.HeadObject(&s3.HeadObjectInput{
		Bucket: aws.String(c.bucket),
		Key:    aws.String(path),
	})

	return
}


resizer.go
package imageresizer

import (
	"errors"
	"fmt"
	_ "golang.org/x/image/bmp"
	_ "golang.org/x/image/tiff"
	_ "golang.org/x/image/webp"
	"image"
	_ "image/gif"
	_ "image/jpeg"
	_ "image/png"
	"math"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
)

type Template struct {
	width   float64
	height  float64
	ratio   float64
	crop    bool
	upscale bool
}

var convertCmd = GetEnv("CMD_CONVERT", "convert")

var optimizerCmd = map[string]string{
	"png":  GetEnv("CMD_OPTIMIZER_PNG", "pngquant --force --ext .png --skip-if-larger --quality 0-75 --speed 4 --strip --") + " %file",
	"jpg":  GetEnv("CMD_OPTIMIZER_JPG", "jpegoptim --force --strip-all --max 75 --quiet --all-progressive") + " %file",
	"gif":  GetEnv("CMD_OPTIMIZER_GIF", "gifsicle --batch --optimize=3") + " %file",
	"svg":  GetEnv("CMD_OPTIMIZER_SVG", "svgcleaner") + " %file %file",
	"webp": GetEnv("CMD_OPTIMIZER_WEBP", "cwebp -m 6 -pass 10 -mt -q 75 -quiet") + " %file -o %file",
}

var optimizerCmdExtMapper = map[string]string{
	".png":  "png",
	".jpg":  "jpg",
	".jpeg": "jpg",
	".gif":  "gif",
	".webp": "webp",
}

var supportedFormats = [...]string{
	".webp",
	".png",
	".svg",
	".jpg",
	".gif",
	".jpeg",
	".bmp",
	".tiff",
	".pdf",
}

func NewTemplate(template string) *Template {
	width := 0.0
	height := 0.0
	ratio := 0.0
	crop := false
	upscale := false

	parts := strings.Split(template, "-")
	for _, part := range parts {
		if part[0] == 'w' {
			width, _ = strconv.ParseFloat(part[1:], 64)
		} else if part[0] == 'h' {
			height, _ = strconv.ParseFloat(part[1:], 64)
		} else if strings.Contains(part, "x") {
			sides := strings.SplitN(part, "x", 2)
			ratioWidth, _ := strconv.ParseFloat(sides[0], 64)
			ratioHeight, _ := strconv.ParseFloat(sides[1], 64)
			if ratioHeight > 0 {
				ratio = ratioWidth / ratioHeight
			}
		} else if part == "crop" {
			crop = true
		} else if part == "upscale" {
			upscale = true
		}
	}

	return &Template{width, height, ratio, crop, upscale}
}

func (t *Template) getDimensions(originalWidth float64, originalHeight float64) (int, int) {
	outputWidth := originalWidth
	outputHeight := originalHeight
	originalRatio := originalWidth / originalHeight

	if t.width > 0 && t.height > 0 {
		outputWidth = t.width
		outputHeight = t.height
	} else if t.width > 0 && t.ratio > 0 {
		outputWidth = t.width
		outputHeight = t.width / t.ratio
	} else if t.height > 0 && t.ratio > 0 {
		outputWidth = t.height * t.ratio
		outputHeight = t.height
	} else if t.ratio > 0 {
		if originalRatio < t.ratio {
			outputWidth = originalHeight * t.ratio
			outputHeight = originalHeight
		} else {
			outputWidth = originalWidth
			outputHeight = originalWidth / t.ratio
		}
	} else if t.width > 0 {
		outputWidth = t.width
		outputHeight = t.width / originalRatio
	} else if t.height > 0 {
		outputWidth = t.height * originalRatio
		outputHeight = t.height
	}

	if t.crop && !t.upscale {
		upscaleWidth := originalWidth / outputWidth
		upscaleHeight := originalHeight / outputHeight
		outputRatio := outputWidth / outputHeight

		if upscaleWidth < 1 && upscaleWidth <= upscaleHeight {
			outputWidth = originalWidth
			outputHeight = originalWidth / outputRatio
		} else if upscaleHeight < 1 && upscaleHeight < upscaleWidth {
			outputWidth = originalHeight * outputRatio
			outputHeight = originalHeight
		}
	}

	return int(math.Round(outputWidth)), int(math.Round(outputHeight))
}

func GetImageDimensions(p string) (w int, h int, err error) {
	f, err := os.Open(p)
	if err != nil {
		return
	}

	c, _, err := image.DecodeConfig(f)
	if err != nil {
		return
	}

	return c.Width, c.Height, nil
}

func ConvertFile(inputFilePath string, inputWidth int, inputHeight int, outputFilePath string, template *Template) (err error) {
	outputWidth, outputHeight := template.getDimensions(float64(inputWidth), float64(inputHeight))
	outputExt := strings.ToLower(filepath.Ext(outputFilePath))
	var resizeGeometryArg, backgroundArg string

	if template.crop {
		resizeGeometryArg = "%dx%d^"
	} else if template.upscale {
		resizeGeometryArg = "%dx%d"
	} else {
		resizeGeometryArg = "%dx%d>"
	}

	if outputExt == ".jpg" || outputExt == ".jpeg" {
		backgroundArg = "white"
	} else {
		backgroundArg = "none"
	}

	args := []string{
		inputFilePath,
		"-resize", fmt.Sprintf(resizeGeometryArg, outputWidth, outputHeight),
		"-background", backgroundArg,
		"-gravity", "center",
		"-extent", fmt.Sprintf("%dx%d", outputWidth, outputHeight),
		outputFilePath,
	}

	_ = os.MkdirAll(filepath.Dir(outputFilePath), 0777)

	logger("RESIZE %s %v\n", convertCmd, args)
	err = exec.Command(convertCmd, args...).Run()

	return
}

func OptimizeFile(filePath string) (err error) {
	ext := strings.ToLower(filepath.Ext(filePath))
	optimizer, found := optimizerCmd[optimizerCmdExtMapper[ext]]
	if !found || optimizer == "" {
		return nil
	}

	args := strings.Split(strings.ReplaceAll(optimizer, "%file", filePath), " ")

	logger("OPTIMIZE %v\n", args)
	err = exec.Command(args[0], args[1:]...).Run()

	return
}

func ValidateFilename(filePath string) error {
	ext := strings.ToLower(filepath.Ext(filePath))

	for _, f := range supportedFormats {
		if f == ext {
			return nil
		}
	}

	return HttpError{errors.New("not supported file format"), http.StatusBadRequest}
}

func ValidateTemplate(template string) error {
	if template != "original" && !strings.HasPrefix(template, "custom") {
		return HttpError{errors.New("not supported template"), http.StatusBadRequest}
	}

	return nil
}

func ShouldNotResize(template string, filePath string) bool {
	if template == "original" {
		return true
	}

	ext := strings.ToLower(filepath.Ext(filePath))
	return ext == ".svg" || ext == ".pdf"
}
